File: MapReduce_FastQ_SequenceQual_Baseline/src/main/java/sequenceQual/FastQ_PerSequenceQual_Mapper.java
package sequenceQual;

import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.lib.input.FileSplit;
import org.seqdoop.hadoop_bam.SequencedFragment;
import utils.IdentifierPositionKeyWritable;

import java.io.IOException;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 04.02.16.
 */
public class FastQ_PerSequenceQual_Mapper extends org.apache.hadoop.mapreduce.Mapper<Object, SequencedFragment, IdentifierPositionKeyWritable, IntWritable> {
    private static final int OFFSET = 33;

    @Override
    protected void map(Object key, SequencedFragment value, Context context) throws IOException, InterruptedException {
        final String sampleIdentifier = ((FileSplit)context.getInputSplit()).getPath().getName();
        context.write(new IdentifierPositionKeyWritable(sampleIdentifier, getMeanValue(value.getQuality())), new IntWritable(1));
    }

    private Integer getMeanValue(Text quality) {
        int sumOfQualityValues = 0;
        for(Character qualVal : quality.toString().toCharArray()) {
            sumOfQualityValues += getCorrespondingIntValue(qualVal);
        }
        return (int)(Math.round((double)sumOfQualityValues/(double)quality.getLength()));
    }

    private int getCorrespondingIntValue(Character qualVal) {
        int intQualVal = (int)qualVal;
        return intQualVal-FastQ_PerSequenceQual_Mapper.OFFSET;
    }

}

File: Spark_FastQ_SequenceQual_Baseline/src/main/scala/FastQ_PerSequenceQual_Mapper.scala
package main.scala

import org.apache.hadoop.io.Text
import org.seqdoop.hadoop_bam.SequencedFragment

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 04.02.16.
  */
object FastQ_PerSequenceQual_Mapper {
  private val OFFSET: Int = 33

  def map(sampleIdentifier:String, record: SequencedFragment): Pair[Pair[String, Int], Int] = {
    new Pair(new Pair(sampleIdentifier, getMeanValue(record.getQuality)), 1)
  }

  private def getMeanValue(quality: Text): Int = {
    var sumOfQualityValues: Int = 0
    for (qualVal <- quality.toString.toCharArray) {
      sumOfQualityValues += getCorrespondingIntValue(qualVal)
    }
    return (Math.round(sumOfQualityValues.toDouble / quality.getLength.toDouble).toInt);
  }

  private def getCorrespondingIntValue(qualVal: Char): Int = {
    qualVal.toInt - OFFSET
  }

}